File: frontend/src/types/asset.ts
export type AssetType =
  | 'host'
  | 'server'
  | 'workstation'
  | 'laptop'
  | 'mobile'
  | 'network_device'
  | 'account'
  | 'service_account'
  | 'file'
  | 'directory'
  | 'url'
  | 'email_address'
  | 'database'
  | 'application'
  | 'cloud_resource'
  | 'other'

export type AssetStatus = 'suspected' | 'confirmed' | 'remediated' | 'cleared'
export type AssetCriticality = 'critical' | 'high' | 'medium' | 'low'

export type AssetLinkType =
  | 'communicates_with'
  | 'owns'
  | 'runs'
  | 'connects_to'
  | 'authenticates_to'
  | 'contains'
  | 'accesses'
  | 'lateral_movement'
  | 'exfiltration_target'
  | 'related'

export interface Asset {
  id: string
  incident_id: string
  asset_type: AssetType
  name: string
  description: string | null
  status: AssetStatus
  criticality: AssetCriticality
  tags: string[]
  metadata: Record<string, unknown>
  added_by: string | null
  created_at: string
  updated_at: string
}

export interface AssetLink {
  id: string
  incident_id: string
  source_id: string
  target_id: string
  link_type: AssetLinkType
  label: string | null
  created_by: string | null
  created_at: string
}

// ── Display helpers ───────────────────────────────────────────────────────────

export const ASSET_TYPE_ICONS: Record<AssetType, string> = {
  host: '🖥️',
  server: '🗄️',
  workstation: '💻',
  laptop: '💻',
  mobile: '📱',
  network_device: '🌐',
  account: '👤',
  service_account: '🤖',
  file: '📄',
  directory: '📁',
  url: '🔗',
  email_address: '📧',
  database: '🗃️',
  application: '⚙️',
  cloud_resource: '☁️',
  other: '📦',
}

export const ASSET_TYPE_LABELS: Record<AssetType, string> = {
  host: 'Host',
  server: 'Server',
  workstation: 'Workstation',
  laptop: 'Laptop',
  mobile: 'Mobile Device',
  network_device: 'Network Device',
  account: 'User Account',
  service_account: 'Service Account',
  file: 'File',
  directory: 'Directory',
  url: 'URL',
  email_address: 'Email Address',
  database: 'Database',
  application: 'Application',
  cloud_resource: 'Cloud Resource',
  other: 'Other',
}

export const ASSET_STATUS_COLORS: Record<AssetStatus, string> = {
  suspected: 'chip-yellow',
  confirmed: 'chip-red',
  remediated: 'chip-green',
  cleared: 'chip-muted',
}

export const ASSET_CRITICALITY_COLORS: Record<AssetCriticality, string> = {
  critical: 'chip-red',
  high: 'chip-yellow',
  medium: 'chip-blue',
  low: 'chip-muted',
}

export const ASSET_LINK_TYPE_LABELS: Record<AssetLinkType, string> = {
  communicates_with: 'Communicates with',
  owns: 'Owns',
  runs: 'Runs',
  connects_to: 'Connects to',
  authenticates_to: 'Authenticates to',
  contains: 'Contains',
  accesses: 'Accesses',
  lateral_movement: 'Lateral movement to',
  exfiltration_target: 'Exfiltration target',
  related: 'Related to',
}

export const ASSET_TYPES_LIST: AssetType[] = [
  'host', 'server', 'workstation', 'laptop', 'mobile', 'network_device',
  'account', 'service_account', 'file', 'directory', 'url', 'email_address',
  'database', 'application', 'cloud_resource', 'other',
]

export const ASSET_LINK_TYPES_LIST: AssetLinkType[] = [
  'communicates_with', 'owns', 'runs', 'connects_to', 'authenticates_to',
  'contains', 'accesses', 'lateral_movement', 'exfiltration_target', 'related',
]

File: backend/app/schemas/asset.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel

ASSET_TYPES = Literal[
    "host",
    "server",
    "workstation",
    "laptop",
    "mobile",
    "network_device",
    "account",
    "service_account",
    "file",
    "directory",
    "url",
    "email_address",
    "database",
    "application",
    "cloud_resource",
    "other",
]

ASSET_STATUSES = Literal["suspected", "confirmed", "remediated", "cleared"]
ASSET_CRITICALITIES = Literal["critical", "high", "medium", "low"]


class AssetCreate(BaseModel):
    asset_type: ASSET_TYPES
    name: str
    description: str | None = None
    status: ASSET_STATUSES = "suspected"
    criticality: ASSET_CRITICALITIES = "medium"
    tags: list[str] = []
    metadata: dict = {}


class AssetBulkCreate(BaseModel):
    """Create multiple assets at once (one per line in UI)."""
    asset_type: ASSET_TYPES
    names: list[str]
    description: str | None = None
    status: ASSET_STATUSES = "suspected"
    criticality: ASSET_CRITICALITIES = "medium"
    tags: list[str] = []


class AssetUpdate(BaseModel):
    asset_type: ASSET_TYPES | None = None
    name: str | None = None
    description: str | None = None
    status: ASSET_STATUSES | None = None
    criticality: ASSET_CRITICALITIES | None = None
    tags: list[str] | None = None
    metadata: dict | None = None


class AssetOut(BaseModel):
    id: UUID
    incident_id: UUID
    asset_type: str
    name: str
    description: str | None
    status: str
    criticality: str
    tags: list[str]
    metadata: dict
    added_by: UUID | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @classmethod
    def model_validate(cls, obj, *args, **kwargs):
        if hasattr(obj, "metadata_"):
            obj.__dict__["metadata"] = obj.metadata_
        return super().model_validate(obj, *args, **kwargs)


class AssetLinkCreate(BaseModel):
    source_id: UUID
    target_id: UUID
    link_type: Literal[
        "communicates_with",
        "owns",
        "runs",
        "connects_to",
        "authenticates_to",
        "contains",
        "accesses",
        "lateral_movement",
        "exfiltration_target",
        "related",
    ]
    label: str | None = None


class AssetLinkOut(BaseModel):
    id: UUID
    incident_id: UUID
    source_id: UUID
    target_id: UUID
    link_type: str
    label: str | None
    created_by: UUID | None
    created_at: datetime

    model_config = {"from_attributes": True}


class AssetTimelineLinkCreate(BaseModel):
    asset_ids: list[UUID]
    timeline_entry_id: UUID
